File: src/Servers/Kestrel/samples/NetworkProtoSample/scripts/verify-wrk.py
import json
import pathlib
import re
import sys

prefix, mode, seconds = sys.argv[1:]
seconds = int(seconds)
text = pathlib.Path(prefix + ".txt").read_text()
elapsed = float(re.search(r"requests in ([\d.]+)s", text)[1])
monotonic = float(pathlib.Path(prefix + ".end-monotonic").read_text()) - float(pathlib.Path(prefix + ".start-monotonic").read_text())
assert seconds * 0.95 <= elapsed <= seconds + 1, (seconds, elapsed)
assert abs(monotonic - elapsed) < 0.5, f"Clock mismatch: wrk={elapsed}s monotonic={monotonic}s"
match = re.search(r"VERIFY (.+)", text)
assert match, "wrk response verification did not run"
counts = {key: int(value) for key, value in re.findall(r"(\w+)=(\d+)", match[1])}
assert counts["responses"] == counts["requests"] > 0, counts
assert all(counts[key] == 0 for key in ("bad", "connect", "read", "write", "status", "timeout")), counts
# wrk can omit stalled requests from both its completed-latency histogram and
# timeout count at shutdown. Reject large gaps between active samples and the
# full-interval rate instead of accepting those counters alone.
active = re.search(r"Req/Sec\s+([\d.]+)([kM]?)", text)
threads = int(re.search(r"(\d+) threads", text)[1])
aggregate = float(re.search(r"Requests/sec:\s+([\d.]+)", text)[1])
active_rate = float(active[1]) * {"": 1, "k": 1000, "M": 1000000}[active[2]] * threads
assert aggregate >= active_rate * 0.9, f"Possible stalled traffic: interval {aggregate} req/s versus active samples {active_rate} req/s"
before = json.loads(pathlib.Path(prefix + ".before.json").read_text())
after = json.loads(pathlib.Path(prefix + ".after.json").read_text())
delta = {key: after[key] - before[key] for key in ("accepted", "closed", "handshakes", "requests", "reusedRequests")}
assert before["scheme"] == after["scheme"]
if after["scheme"] == "http":
    assert delta["handshakes"] == 0, delta
else:
    assert delta["handshakes"] > 0, delta
assert delta["requests"] >= counts["responses"], delta
if mode == "close":
    assert counts["reused"] == delta["reusedRequests"] == 0, (counts, delta)
    assert delta["accepted"] >= delta["requests"], delta
    if after["scheme"] == "https":
        assert delta["handshakes"] >= delta["requests"], delta
else:
    assert counts["reused"] > 0 and delta["reusedRequests"] > 0, (counts, delta)
    assert delta["requests"] > 10 * delta["accepted"], delta
print(f"{mode} verified: {counts}; server delta={delta}")
if "allocatedBytes" in after:
    allocated = after["allocatedBytes"] - before["allocatedBytes"]
    cpu = after["cpuMilliseconds"] - before["cpuMilliseconds"]
    print(f"Server allocation={allocated / delta['requests']:.1f} B/request; CPU={cpu * 1000 / delta['requests']:.2f} us/request; GC counts={[after[f'gen{i}'] - before[f'gen{i}'] for i in range(3)]}")
print("Metrics probes add one connection; requests completed by server at wrk cutoff may exceed client responses.")
print("Plaintext: no TLS." if after["scheme"] == "http" else "TLS session resumption unverified: fresh TCP connections are NOT asserted to be full certificate handshakes.")

File: src/Servers/Kestrel/samples/NetworkProtoSample/scripts/verify-wrk.lua
local threads = {}
local expected_body = string.rep("x", 1023) .. "\n"

function setup(thread)
    table.insert(threads, thread)
end

function init(args)
    mode = args[1]
    backend = args[2]
    scheme = args[3]
    bad = 0
    seen = 0
    reused = 0
    if mode == "close" then
        wrk.headers["Connection"] = "close"
    end
end

function response(status, headers, body)
    seen = seen + 1
    local number = tonumber(headers["X-Connection-Request"])
    local id = headers["X-Connection-Id"] or ""
    if number and number > 1 then reused = reused + 1 end
    if status ~= 200 or body ~= expected_body or headers["X-Backend"] ~= backend
        or headers["X-Tls"] ~= (scheme == "http" and "none" or "Tls12:TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256")
        or not number or ((string.sub(id, 1, 9) == "io-uring-") ~= (backend == "io_uring"))
        or (mode == "close" and (number ~= 1 or headers["Connection"] ~= "close")) then
        bad = bad + 1
    end
end

function done(summary, latency, requests)
    local count, failures, reuse = 0, 0, 0
    for _, thread in ipairs(threads) do
        count = count + thread:get("seen")
        failures = failures + thread:get("bad")
        reuse = reuse + thread:get("reused")
    end
    print(string.format("VERIFY responses=%d bad=%d reused=%d requests=%d connect=%d read=%d write=%d status=%d timeout=%d",
        count, failures, reuse, summary.requests, summary.errors.connect, summary.errors.read,
        summary.errors.write, summary.errors.status, summary.errors.timeout))
end
